pkg/migration/migrations/public_schema_migration.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package migrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/migration"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkeys"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/dbdesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/schemadesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/errors"
+)
+
+func publicSchemaMigration(
+	ctx context.Context, _ clusterversion.ClusterVersion, d migration.TenantDeps, _ *jobs.Job,
+) error {
+	query := `
+  SELECT ns_db.id
+    FROM system.namespace AS ns_db
+         INNER JOIN system.namespace AS ns_sc ON (
+                                                                ns_db.id
+                                                                = ns_sc."parentID"
+                                                             )
+   WHERE ns_db.id != 1
+     AND ns_db."parentSchemaID" = 0
+     AND ns_db."parentID" = 0
+     AND ns_sc."parentSchemaID" = 0
+     AND ns_sc.name = 'public'
+     AND ns_sc.id = 29
+ORDER BY ns_db.id ASC;
+`
+	rows, err := d.InternalExecutor.QueryIterator(
+		ctx, "get_databases_with_synthetic_public_schemas", nil /* txn */, query,
+	)
+	if err != nil {
+		return err
+	}
+	var databaseIDs []descpb.ID
+	for ok, err := rows.Next(ctx); ok; ok, err = rows.Next(ctx) {
+		if err != nil {
+			return err
+		}
+		parentID := descpb.ID(tree.MustBeDInt(rows.Cur()[0]))
+		databaseIDs = append(databaseIDs, parentID)
+	}
+
+	for _, dbID := range databaseIDs {
+		if err := createPublicSchemaForDatabase(ctx, dbID, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createPublicSchemaForDatabase(
+	ctx context.Context, dbID descpb.ID, d migration.TenantDeps,
+) error {
+	return d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
+	) error {
+
+		found, desc, err := descriptors.GetImmutableDatabaseByID(ctx, txn, dbID, tree.DatabaseLookupFlags{})
```

Comment:
One transaction per database is the way to go, but wouldn't it be safer to also check here also that `desc`'s public schema is indeed descriptorless, and return early if for some reason it now has one? 

Bear in mind that this transaction is not the same as the one that ran your namespace join (and that's fine). I'm not sure if this matters, but even so you're better off checking anyway, because you got this db id from the namespace table, not because you directly know for a fact that this db has no public schema descriptor. It should be equivalent, but who knows if that's always the case. Let's just stay on the safe side and check.